### PR TITLE
A prototype for mgmt generation report

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes https://github.com/Azure/autorest.csharp/issues/1647

# Description
<i>Add your description here!</i>

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/Mgmt/Decorator/CodeModelExtension.cs [hunk 1]
```diff
@@ -167,11 +186,15 @@ public static void UpdateSubscriptionIdForAllResource(this CodeModel codeModel)
...
                             p.Implementation = ImplementationLocation.Method;
+                            // TODO -- change the operation name to `op.OperationId` when it is ready
+                            MgmtContext.Report.Add(ReportLevel.Information, $"The implementation location of parameter {p.Language.Default.Name} in operation `{operationGroup.Key}_{op.Language.Default.Name}` has been changed to Method");
                         }
                         // update the apiVersion parameter to be 'client' method
                         if (p.Language.Default.Name.Equals("apiVersion", StringComparison.OrdinalIgnoreCase))
                         {
                             p.Implementation = ImplementationLocation.Client;
+                            // TODO -- change the operation name to `op.OperationId` when it is ready
+                            MgmtContext.Report.Add(ReportLevel.Information, $"The implementation location of parameter {p.Language.Default.Name} in operation `{operationGroup.Key}_{op.Language.Default.Name}` has been changed to Client");
```
Comment (fengzhou-msft) [inline src/AutoRest.CSharp/Mgmt/Decorator/CodeModelExtension.cs:197]: The original location for apiVersion is Client for most cases, only in a few cases like [ResourcesRestOperations](https://github.com/Azure/azure-sdk-for-net/blob/1e11c266329079133cdb62c7fe9738068002c529/sdk/resourcemanager/Azure.ResourceManager/src/Resources/Custom/RestOperations/ResourcesRestOperations.cs#L253), the location was Method (The customized code actually moves back the location for apiVersion to Method to offset the change here. There should be a better example for why the change is made here, but I cannot remember now).

Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Mgmt/Decorator/CodeModelExtension.cs:197]: GenericResource CRUD need the api-version to be a method parameter, I believe this should be the only use case for this.

I will add a if-check here only to output the log line when there is a change.
Or I could introduce a method in the `Report` class, to determine whether we need the output or not.

**Comment by fengzhou-msft** [comment]
Shall we have report for common type replacement?

**Comment by ArcturusZhang** [comment]
> Shall we have report for common type replacement?

I believe we should have everything that we have changed (comparing with the modeler) in the report, therefore this should be included.

**Pull request closed without merging**